dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/event/JdbcPredicate.java

```diff
@@ -0,0 +1,230 @@
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.springframework.jdbc.core.SqlParameterValue;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+
+/**
+ * JdbcPredicate turns a {@link QueryFilter} on a data element or tracked entity attribute into a
+ * SQL predicate with an optional JDBC parameter (right-hand side operand of the SQL operator). The
+ * {@link #sql()} can be inserted into a SQL statement where clause. User input in the {@link
+ * QueryFilter#getFilter()} is transformed into the {@link #parameter()}. Non-unary operators will
+ * have a parameter which must be added to the JDBC template parameter source like {@link
+ * MapSqlParameterSource}!
+ *
+ * <p>These are some of the transformations we need to do on user input
+ *
+ * <ul>
+ *   <li>split input {@code "one;two"} for operator {@link QueryOperator#IN}
+ *   <li>convert input value(s) from text to other data types so comparisons can be done using DHIS2
+ *       data value type semantics i.e. {@code 170.0 == 170} which is what users expect for value
+ *       type integer/number while {@code "170.0" != "170"} is unexpected but what happens for
+ *       treating the values as text
+ *   <li>escape wildcard characters '%' and '_' for operators backed by SQL {@code like}
+ *   <li>insert '%' wildcard(s) before/after user input for like based operators
+ * </ul>
+ *
+ * Other special characters like single quotes are expected to be escaped by the <a
+ * href="https://github.com/pgjdbc/pgjdbc/blob/156d724e1d95052b41a19fb568b2f81919ae2197/pgjdbc/src/main/java/org/postgresql/core/v3/SimpleParameterList.java#L213">JDBC
+ * driver</a>!
+ *
+ * <p>On {@link SqlParameterValue}: Spring handles {@code List.of()} values for us no matter if we
+ * pass it to an <a
+ * href="https://docs.spring.io/spring-framework/reference/data-access/jdbc/parameter-handling.html#jdbc-in-clause">in</a>
+ * operator allowing multiple values or an operator only expecting one like {@code eq}.
+ * JdbcPredicate ensures we never pass a collection with multiple elements to such operators as this
+ * leads to a {@code BadSqlGrammarException}.
+ */
+record JdbcPredicate(String sql, Optional<Parameter> parameter) {
+  record Parameter(String name, SqlParameterValue value) {}
+
+  public static JdbcPredicate of(@Nonnull TrackedEntityAttribute tea, @Nonnull QueryFilter filter) {
+    Parameter parameter = parseFilterValue(tea, filter);
+    String sql = generateSql(tea, filter, parameter);
+    return new JdbcPredicate(sql, Optional.ofNullable(parameter));
+  }
+
+  public static JdbcPredicate of(
+      @Nonnull DataElement de, @Nonnull QueryFilter filter, @Nonnull String tableName) {
+    Parameter parameter = parseFilterValue(de, filter);
+    String sql = generateSql(de, filter, parameter, tableName);
+    return new JdbcPredicate(sql, Optional.ofNullable(parameter));
+  }
+
+  @SuppressWarnings("unchecked")
+  private static Parameter parseFilterValue(
```

Comment:
I am not really fond of `Optional` fields in object but I can see the point here.
Can we at least make this method return a `Optional<Parameter>` so we avoid to call `Optional.ofNullable` in the caller?

Reply:
If I do that I get linting complaints 😂 I don't mind as its there to avoid null, providing a different API. I don't really see why it would be odd. Either you deal with null or with Optional 🤷🏻 (or you make subclasses or so for Unary, Binary which seems too much)

Reply:
I would prefer the linting complaint over the use of `Optional.ofNullable`.
Just saying that because we already know from the `parseFilterValue` if the parameter is there or not.